From 38a18a447880643964e6e352d1a364237830815e Mon Sep 17 00:00:00 2001
From: Kjell Ahlstedt <kjellahlstedt@gmail.com>
Date: Tue, 7 Jul 2026 09:49:06 +0200
Subject: [PATCH] Gdk: Don't derive gtkmm__GdkXxx GTypes from internal types

Add _DO_NOT_DERIVE_GTYPE and _ABI_AS_WITH_DERIVED_GTYPE where the
underlying C class is declared G_DECLARE_FINAL_TYPE or
GDK__DECLARE_INTERNAL_TYPE.
Such declarations have been added to many GdkXxx classes.
See gtk!10116
---
 gdk/src/applaunchcontext.hg | 2 ++
 gdk/src/cairocontext.hg     | 2 ++
 gdk/src/clipboard.hg        | 2 ++
 gdk/src/cursor.hg           | 3 ++-
 gdk/src/device.hg           | 3 +++
 gdk/src/devicepad.hg        | 2 ++
 gdk/src/devicetool.hg       | 2 ++
 gdk/src/display.hg          | 2 ++
 gdk/src/displaymanager.hg   | 2 ++
 gdk/src/dmabuftexture.hg    | 2 ++
 gdk/src/drag.hg             | 2 ++
 gdk/src/drawcontext.hg      | 2 ++
 gdk/src/drop.hg             | 2 ++
 gdk/src/frameclock.hg       | 2 ++
 gdk/src/glcontext.hg        | 2 ++
 gdk/src/gltexture.hg        | 2 ++
 gdk/src/memorytexture.hg    | 2 ++
 gdk/src/monitor.hg          | 2 ++
 gdk/src/seat.hg             | 2 ++
 gdk/src/snapshot.hg         | 2 ++
 gdk/src/surface.hg          | 3 +++
 gdk/src/texture.hg          | 2 ++
 22 files changed, 46 insertions(+), 1 deletion(-)

diff --git a/gdk/src/applaunchcontext.hg b/gdk/src/applaunchcontext.hg
index cc2cec7d..1c5e8e6b 100644
--- a/gdk/src/applaunchcontext.hg
+++ b/gdk/src/applaunchcontext.hg
@@ -35,6 +35,8 @@ class GDKMM_API Display;
 class GDKMM_API AppLaunchContext : public Gio::AppLaunchContext
 {
   _CLASS_GOBJECT(AppLaunchContext, GdkAppLaunchContext, GDK_APP_LAUNCH_CONTEXT, Gio::AppLaunchContext, GAppLaunchContext, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkAppLaunchContext is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
 
 protected:
   _CTOR_DEFAULT()
diff --git a/gdk/src/cairocontext.hg b/gdk/src/cairocontext.hg
index 46087539..730838b5 100644
--- a/gdk/src/cairocontext.hg
+++ b/gdk/src/cairocontext.hg
@@ -36,6 +36,8 @@ namespace Gdk
 class GDKMM_API CairoContext : public Gdk::DrawContext
 {
   _CLASS_GOBJECT(CairoContext, GdkCairoContext, GDK_CAIRO_CONTEXT, Gdk::DrawContext, GdkDrawContext, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkCairoContext is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
 
 protected:
   _CTOR_DEFAULT()
diff --git a/gdk/src/clipboard.hg b/gdk/src/clipboard.hg
index 54f9555d..b0f68be4 100644
--- a/gdk/src/clipboard.hg
+++ b/gdk/src/clipboard.hg
@@ -56,6 +56,8 @@ namespace Gdk
 class GDKMM_API Clipboard : public Glib::Object
 {
   _CLASS_GOBJECT(Clipboard, GdkClipboard, GDK_CLIPBOARD, Glib::Object, GObject, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkClipboard is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
 
 public:
   _WRAP_METHOD(Glib::RefPtr<Display> get_display(), gdk_clipboard_get_display, refreturn)
diff --git a/gdk/src/cursor.hg b/gdk/src/cursor.hg
index 05cb8882..da3137c6 100644
--- a/gdk/src/cursor.hg
+++ b/gdk/src/cursor.hg
@@ -60,7 +60,8 @@ namespace Gdk
 class GDKMM_API Cursor : public Glib::Object
 {
   _CLASS_GOBJECT(Cursor, GdkCursor, GDK_CURSOR, Glib::Object, GObject, , , GDKMM_API)
-  _DO_NOT_DERIVE_GTYPE //We use DO_NOT_DERIVE_GTYPE because glib does not allow us to derive from a non-fundamental (abstract) type, for some reason.
+  _DO_NOT_DERIVE_GTYPE dnl// GdkCursor is a final type
+  _STRUCT_NOT_HIDDEN
 
 protected:
 
diff --git a/gdk/src/device.hg b/gdk/src/device.hg
index d5e49190..6fc86ab6 100644
--- a/gdk/src/device.hg
+++ b/gdk/src/device.hg
@@ -39,6 +39,9 @@ class GDKMM_API Surface;
 class GDKMM_API Device : public Glib::Object
 {
   _CLASS_GOBJECT(Device,GdkDevice,GDK_DEVICE,Glib::Object,GObject, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkDevice is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
+
   _CUSTOM_WRAP_NEW
 
 protected:
diff --git a/gdk/src/devicepad.hg b/gdk/src/devicepad.hg
index b88bb660..35942523 100644
--- a/gdk/src/devicepad.hg
+++ b/gdk/src/devicepad.hg
@@ -56,6 +56,8 @@ namespace Gdk
 class GDKMM_API DevicePad : public Glib::Interface
 {
   _CLASS_INTERFACE(DevicePad, GdkDevicePad, GDK_DEVICE_PAD, GdkDevicePadInterface, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkDevicePad is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
 
 public:
   _WRAP_ENUM(Feature, GdkDevicePadFeature, decl_prefix GDKMM_API)
diff --git a/gdk/src/devicetool.hg b/gdk/src/devicetool.hg
index 3b51d45a..663038ae 100644
--- a/gdk/src/devicetool.hg
+++ b/gdk/src/devicetool.hg
@@ -31,6 +31,8 @@ namespace Gdk
 class GDKMM_API DeviceTool : public Glib::Object
 {
   _CLASS_GOBJECT(DeviceTool, GdkDeviceTool, GDK_DEVICE_TOOL, Glib::Object, GObject, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkDeviceTool is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
 
 public:
   _WRAP_ENUM(Type, GdkDeviceToolType, decl_prefix GDKMM_API)
diff --git a/gdk/src/display.hg b/gdk/src/display.hg
index b3f9ac95..a386f8af 100644
--- a/gdk/src/display.hg
+++ b/gdk/src/display.hg
@@ -52,6 +52,8 @@ class GDKMM_API Surface;
 class GDKMM_API Display : public Glib::Object
 {
   _CLASS_GOBJECT(Display,GdkDisplay,GDK_DISPLAY,Glib::Object,GObject, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkDisplay is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
 protected:
   _CTOR_DEFAULT
 
diff --git a/gdk/src/displaymanager.hg b/gdk/src/displaymanager.hg
index 939b3f0c..70b456cb 100644
--- a/gdk/src/displaymanager.hg
+++ b/gdk/src/displaymanager.hg
@@ -32,6 +32,8 @@ namespace Gdk
 class GDKMM_API DisplayManager : public Glib::Object
 {
   _CLASS_GOBJECT(DisplayManager,GdkDisplayManager,GDK_DISPLAY_MANAGER,Glib::Object,GObject, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkDisplayManager is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
 protected:
 
 public:
diff --git a/gdk/src/dmabuftexture.hg b/gdk/src/dmabuftexture.hg
index 94ba74d2..5a823e32 100644
--- a/gdk/src/dmabuftexture.hg
+++ b/gdk/src/dmabuftexture.hg
@@ -36,6 +36,8 @@ _WRAP_GERROR(DmabufError, GdkDmabufError, GDK_DMABUF_ERROR, decl_prefix GDKMM_AP
 class GDKMM_API DmabufTexture : public Gdk::Texture
 {
   _CLASS_GOBJECT(DmabufTexture, GdkDmabufTexture, GDK_DMABUF_TEXTURE, Gdk::Texture, GdkTexture, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkDmabufTexture is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
 
 protected:
   _CTOR_DEFAULT()
diff --git a/gdk/src/drag.hg b/gdk/src/drag.hg
index ffa194ab..4540cddf 100644
--- a/gdk/src/drag.hg
+++ b/gdk/src/drag.hg
@@ -44,6 +44,8 @@ _WRAP_ENUM(DragCancelReason, GdkDragCancelReason, decl_prefix GDKMM_API)
 class GDKMM_API Drag : public Glib::Object
 {
   _CLASS_GOBJECT(Drag, GdkDrag, GDK_DRAG, Glib::Object, GObject, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkDrag is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
 
 public:
   // The Gtk::DnD namespace contains functions that logically are Gdk::Drag methods.
diff --git a/gdk/src/drawcontext.hg b/gdk/src/drawcontext.hg
index 13ab7a49..bb3d5682 100644
--- a/gdk/src/drawcontext.hg
+++ b/gdk/src/drawcontext.hg
@@ -40,6 +40,8 @@ class GDKMM_API Surface;
 class GDKMM_API DrawContext : public Glib::Object
 {
   _CLASS_GOBJECT(DrawContext, GdkDrawContext, GDK_DRAW_CONTEXT, Glib::Object, GObject, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkDrawContext is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
 
 protected:
   /// This constructor is protected because only derived classes shall be instantiated.
diff --git a/gdk/src/drop.hg b/gdk/src/drop.hg
index 40cb0ad5..e02898f3 100644
--- a/gdk/src/drop.hg
+++ b/gdk/src/drop.hg
@@ -55,6 +55,8 @@ class GDKMM_API Surface;
 class GDKMM_API Drop : public Glib::Object
 {
   _CLASS_GOBJECT(Drop, GdkDrop, GDK_DROP, Glib::Object, GObject, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkDrop is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
 
 public:
   _WRAP_METHOD(Glib::RefPtr<Display> get_display(), gdk_drop_get_display, refreturn)
diff --git a/gdk/src/frameclock.hg b/gdk/src/frameclock.hg
index 70b93b9d..c6fd9603 100644
--- a/gdk/src/frameclock.hg
+++ b/gdk/src/frameclock.hg
@@ -67,6 +67,8 @@ namespace Gdk
 class GDKMM_API FrameClock : public Glib::Object
 {
   _CLASS_GOBJECT(FrameClock, GdkFrameClock, GDK_FRAME_CLOCK, Glib::Object, GObject, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkFrameClock is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
 
 public:
   _WRAP_ENUM(Phase, GdkFrameClockPhase, decl_prefix GDKMM_API)
diff --git a/gdk/src/glcontext.hg b/gdk/src/glcontext.hg
index dd9a2ffe..ea05ddc0 100644
--- a/gdk/src/glcontext.hg
+++ b/gdk/src/glcontext.hg
@@ -79,6 +79,8 @@ _WRAP_GERROR(GLError, GdkGLError, GDK_GL_ERROR, newin "3,18", decl_prefix GDKMM_
 class GDKMM_API GLContext : public DrawContext
 {
   _CLASS_GOBJECT(GLContext, GdkGLContext, GDK_GL_CONTEXT, Gdk::DrawContext, GdkDrawContext, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkGLContext is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
 
 protected:
   _CTOR_DEFAULT
diff --git a/gdk/src/gltexture.hg b/gdk/src/gltexture.hg
index 61ce55ef..c963a431 100644
--- a/gdk/src/gltexture.hg
+++ b/gdk/src/gltexture.hg
@@ -30,6 +30,8 @@ class GDKMM_API GLContext;
 class GDKMM_API GLTexture : public Gdk::Texture
 {
   _CLASS_GOBJECT(GLTexture, GdkGLTexture, GDK_GL_TEXTURE, Gdk::Texture, GdkTexture, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkGLTexture is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
 
 protected:
   _CTOR_DEFAULT()
diff --git a/gdk/src/memorytexture.hg b/gdk/src/memorytexture.hg
index 5b35f91e..c7af876e 100644
--- a/gdk/src/memorytexture.hg
+++ b/gdk/src/memorytexture.hg
@@ -30,6 +30,8 @@ namespace Gdk
 class GDKMM_API MemoryTexture : public Gdk::Texture
 {
   _CLASS_GOBJECT(MemoryTexture, GdkMemoryTexture, GDK_MEMORY_TEXTURE, Gdk::Texture, GdkTexture, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkMemoryTexture is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
 
 protected:
   _CTOR_DEFAULT()
diff --git a/gdk/src/monitor.hg b/gdk/src/monitor.hg
index 50391ee5..4474df38 100644
--- a/gdk/src/monitor.hg
+++ b/gdk/src/monitor.hg
@@ -38,6 +38,8 @@ class GDKMM_API Display;
 class GDKMM_API Monitor : public Glib::Object
 {
   _CLASS_GOBJECT(Monitor, GdkMonitor, GDK_MONITOR, Glib::Object, GObject, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkMonitor is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
 
 protected:
   _CTOR_DEFAULT()
diff --git a/gdk/src/seat.hg b/gdk/src/seat.hg
index 096a67c1..b9a8fe1b 100644
--- a/gdk/src/seat.hg
+++ b/gdk/src/seat.hg
@@ -37,6 +37,8 @@ class GDKMM_API Surface;
 class GDKMM_API Seat : public Glib::Object
 {
   _CLASS_GOBJECT(Seat, GdkSeat, GDK_SEAT, Glib::Object, GObject, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkSeat is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
 
 protected:
   _CTOR_DEFAULT()
diff --git a/gdk/src/snapshot.hg b/gdk/src/snapshot.hg
index e4ca180b..bd925df1 100644
--- a/gdk/src/snapshot.hg
+++ b/gdk/src/snapshot.hg
@@ -33,6 +33,8 @@ namespace Gdk
 class GDKMM_API Snapshot : public Glib::Object
 {
   _CLASS_GOBJECT(Snapshot, GdkSnapshot, GDK_SNAPSHOT, Glib::Object, GObject, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkSnapshot is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
 
 public:
   // This class is almost empty
diff --git a/gdk/src/surface.hg b/gdk/src/surface.hg
index 4453cdb8..d8fbea3b 100644
--- a/gdk/src/surface.hg
+++ b/gdk/src/surface.hg
@@ -56,7 +56,10 @@ class GDKMM_API Monitor;
 class GDKMM_API Surface : public Glib::Object
 {
   _CLASS_GOBJECT(Surface, GdkSurface, GDK_SURFACE, Glib::Object, GObject, , , GDKMM_API)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkSurface is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
   _STRUCT_NOT_HIDDEN
+
   _CUSTOM_WRAP_NEW
   _IGNORE(gdk_surface_destroy, gdk_surface_is_destroyed)
 
diff --git a/gdk/src/texture.hg b/gdk/src/texture.hg
index 1ca4d3f0..2d2c6560 100644
--- a/gdk/src/texture.hg
+++ b/gdk/src/texture.hg
@@ -59,6 +59,8 @@ class GDKMM_API Texture : public Glib::Object, public Paintable
 {
   _CLASS_GOBJECT(Texture, GdkTexture, GDK_TEXTURE, Glib::Object, GObject, , , GDKMM_API)
   _IMPLEMENTS_INTERFACE(Paintable)
+  _DO_NOT_DERIVE_GTYPE dnl// GdkTexture is an internal type
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
   _IGNORE(gdk_texture_downloader_new) dnl// gmmproc thinks this is a Texture method
 
 protected:
-- 
GitLab

